pkg/gql/notifications/broker.go

```diff
@@ -0,0 +1,205 @@
+// blocked on mutex-lock, chan push or any I/O operation.
+// It should not expose anything than a channel.
+
+type Broker struct {
+	id uint
+
+	// active clients subscribed for updates
+	clients *list.List
+	// max number of clients per a broker instance
+	maxClientsCount uint
+
+	// ConnectionChan is a shared queue to buffer incoming websocket connections
+	// closing connChan will terminate the broker
+	ConnectionChan chan wsConn
+
+	// Events
+	eventBus          eventbus.Broker
+	acceptedBlockChan chan block.Block
+	acceptedBlockId   uint32
+}
+
+func NewBroker(id uint, eventBus eventbus.Broker, maxClientsCount uint, connChan chan wsConn) *Broker {
+
+	b := new(Broker)
+	b.eventBus = eventBus
+	b.ConnectionChan = connChan
+	b.acceptedBlockChan, b.acceptedBlockId = consensus.InitAcceptedBlockUpdate(eventBus)
+	b.clients = list.New()
+	b.maxClientsCount = maxClientsCount
+	b.id = id
+	return b
+}
+
+func (b *Broker) Run() {
+
+	// Teardown procedure for the Broker
+	defer func() {
+
+		// It's advisable to reset the broker state entirely.
+		// This would allow later to restart a broker without any leaks.
+		log.Infof("Terminating broker %d", b.id)
+
+		// Unsubscribe from all eventBus events
+		b.eventBus.Unsubscribe(topics.AcceptedBlock, b.acceptedBlockId)
+
+		// Terminate all clients goroutines
+		for e := b.clients.Front(); e != nil; e = e.Next() {
+			c := e.Value.(*wsClient)
+			close(c.msgChan)
+		}
+
+		// reset clients list
+		b.clients.Init()
+
+	}()
+
+	for {
+		// Any of the handlers must be capable of recovering from panic
+		select {
+		// new client connection from webserver
+		case conn, isOpen := <-b.ConnectionChan:
+			if isOpen {
+				b.handleConn(conn)
+			} else {
+				// Terminate the broker when the shared connChan is closed
+				return
+			}
+		// new accepted block from node
+		case blk := <-b.acceptedBlockChan:
+			b.handleBlock(blk)
+		case <-time.After(30 * time.Second):
+			b.handleIdle()
+		}
+	}
+}
+
+// handleBlock handles the topics.AcceptedBlock event emitted from node layer.
+// It packs a json from the block and broadcast it to all active clients
+func (b *Broker) handleBlock(blk block.Block) {
+
+	defer func() {
+		if r := recover(); r != nil {
+			log.Errorf("handleBlock recovered from err: %v", r)
+		}
+	}()
+
+	b.reap()
+
+	msg, err := MarshalBlockMsg(blk)
+	if err != nil {
+		log.Errorf("encoding err: %v", err)
+	}
+
+	b.broadcastMessage(msg)
+}
+
+// handleConn handles a new websocket conn pushed from webserver layer It stores
+// the conn to list of active clients
+func (b *Broker) handleConn(conn wsConn) {
+
+	defer func() {
+		if r := recover(); r != nil {
+			log.Errorf("handleConn recovered from err: %v", r)
+		}
+	}()
+
+	log.Tracef("handleConn %s", conn.RemoteAddr().String())
+
+	b.reap()
+
+	if b.clients.Len() >= int(b.maxClientsCount) {
+		log.Errorf("Too many connections %d", b.clients.Len())
+
+		// Free a slot by removing the oldest connection
+		e := b.clients.Front()
```

Comment:
Maybe it would be a bit of over-engineering, however, in order to prevent force-dropping connections, oldest connections should be dropped only when at full capacity, after reaching a maturity threshold (connections older than X minutes). Additionally, incoming connections, when at full capacity, should enqueue with a random delay to prevent spamming. This would be good as a separate issue

Reply:
The limiter middleware kinda takes care of this already. Please ignore